### PR TITLE
switch from jest to vitest for faster, simpler tests

### PR DESCRIPTION
**Pull request by sqs**
Replaces our usage of jest with vitest. Also removes the babel transpiler. This simplifies our test configuration by a lot, makes tests run 10% faster, and makes further modernizations to our build/test stuff possible (such as using vite for frontend builds).

This removes some of the junit exporting for Buildkite, and the vitest bazel defs don't really cleanly implement bazel testing guidelines (like sharding). But vitest is only used for unit tests (all integration/e2e/regression tests have always run in mocha), so none of them are very slow anyway.

## Test plan

CI

### CODE REVIEW

**Comment by sqs** [comment]
The speed is actually the same as jest to run the whole test suite, but vitest's UI feels much faster and has faster init times. Kinda surprised me that it wasn't actually faster when I timed it. Still, I think this change is worth it for the massive simplification in config (no jest.config.base.js, no transpilation of test stuff, no babel, etc.).

File: dev/defs.bzl [hunk 1]
```diff
@@ -126,11 +126,34 @@ def npm_package(name, srcs = [], **kwargs):
...
+            "--config=$(location :%s)" % vitest_config,
+        ],
+        data = data + native.glob(["**/__fixtures__/**/*"]) + [
+            ":%s" % vitest_config,
+            "//:node_modules/happy-dom",
+            "//:node_modules/jsdom",
+        ],
+        env = {"BAZEL": "1", "CI": "1"},
+        patch_node_fs = False,
+        tags = kwargs.pop("tags", []) + ["no-sandbox"],
```
Comment (valerybugakov) [inline dev/defs.bzl:156]: Using a sandbox for it causes failure? It was the case for `jest`.

**Review by valerybugakov — APPROVED** [review]
Huge!

@jhchabran, with this change, we might be able to run `//client/web:test` with other targets: 
https://github.com/sourcegraph/sourcegraph/blob/3475abf565c111aec236a326cb09ac092fa33dce/dev/ci/internal/ci/bazel_operations.go#L22-L24

File: dev/defs.bzl (hunk 1 above)
Comment (sqs) [reply dev/defs.bzl:157]: Yeah

File: vitest.shared.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        testTimeout: 5000,
+        hookTimeout: 1000,
+        experimentalVmThreads: true,
+        include: [`**/*.test.${TS_EXT}?(x)`],
+        exclude: [
+            '**/integration-test',
+            '**/integration',
+            '**/end-to-end',
+            '**/regression',
+            '**/node_modules',
```
[the remark this answers is not in the document]
Comment (sqs) [reply vitest.shared.ts:27]: I would worry that we would accidentally omit tests from CI if we did that.

File: vitest.workspace.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    packages: string[]
+}
+const workspaceFile = load(readFileSync(path.join(__dirname, 'pnpm-workspace.yaml'), 'utf8')) as PnpmWorkspaceFile
+workspaceFile.packages.push('client/web/dev') // is a tsconfig project but not a pnpm workspace package
+const projectRoots = filterProjectRootsByArgv(
+    workspaceFile.packages.flatMap(p => glob.sync(`${p}/`, { cwd: __dirname })).map(p => p.replace(/\/$/, ''))
+).filter(dir => existsSync(path.join(dir, 'vitest.config.ts')))
+
+// If a project root is given as a command-line argument (such as in `pnpm exec vitest client/web`),
+// then only load that project root, to speed up init time (from ~3.2s to ~1.9s).
```
[the remark this answers is not in the document]
Comment (sqs) [reply vitest.workspace.ts]: This ended up not being a perf improvement. Running `pnpm exec vitest` or `pnpm test` in a `client/*` subdir is still a lot faster than running it from the root.

**Pull request merged**